### PR TITLE
MONGOID-5122 Mongoize embedded fields in the criteria selector

### PR DESCRIPTION
**Pull request by Neilshweky**

### CODE REVIEW

File: spec/integration/matcher_examples_spec.rb [hunk 1]
```diff
@@ -687,10 +689,8 @@
           band.records.where(producers: 100..200).first
         end
 
-        it 'raises InvalidQuery' do
-          lambda do
-            found_record
-          end.should raise_error(Mongoid::Errors::InvalidQuery)
+        it 'finds' do
+          expect(found_record).to eq(record)
```
Comment (Neilshweky) [inline spec/integration/matcher_examples_spec.rb:693]: Now, this definitely changes the outcome of the test, but this looks ultimately the same as the test on 664 and it feels like it should return the correct record. This seems like an improvement, but I don't know a lot about the matcher code so @p-mongo let me know if this is incorrect.

Note: this does not fix MONGOID-4911.

**Comment by johnnyshields** [comment]
Great work Neil you're really on a roll with these fixes. It's much appreciated.

File: lib/mongoid/criteria/queryable.rb [hunk 2]
```diff
@@ -60,10 +60,10 @@ def ==(other)
       # @param [ Hash ] aliases The optional field aliases.
       # @param [ Hash ] serializers The optional field serializers.
       # @param [ Symbol ] driver The driver being used.
-      def initialize(aliases = {}, serializers = {}, driver = :mongo)
+      def initialize(aliases = {}, serializers = {}, relations = {}, driver = :mongo)
```
[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongoid/criteria/queryable.rb:63]: I filed https://jira.mongodb.org/browse/MONGOID-5267 for driver argument removal.

File: spec/integration/matcher_examples_spec.rb (hunk 1 above)
Comment (p-mongo) [reply spec/integration/matcher_examples_spec.rb:693]: This test uses simple values that aren't subject to MONGOID-4911 and therefore it does pass. The change you've made in this PR is to apply existing Range expansion to embedded matching, which I think is fine if we are going to then work on MONGOID-4911 and repair how it works.

I would leave the current test as you have it but add a test for `band.records.where(producers: 199..200).first` which would find the band currently but should return nil, and mark it pending on MONGOID-4911.

File: lib/mongoid/criteria/queryable/smash.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@ class Smash < Hash
 
         # @attribute [r] aliases The aliases.
         # @attribute [r] serializers The serializers.
-        attr_reader :aliases, :serializers
+        attr_reader :aliases, :serializers, :relations
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongoid/criteria/queryable/smash.rb:12]: done

File: lib/mongoid/criteria/queryable/selector.rb [hunk 4]
```diff
@@ -182,6 +184,21 @@ def evolve_hash(serializer, value)
...
 
+        # Evolve a single key selection with range values.
+        #
+        # @api private
+        #
+        # @param [ Object ] serializer The optional serializer for the field.
+        # @param [ Range ] value The Range to serialize.
+        #
+        # @return [ Range ] The serialized Range.
+        def evolve_range(serializer, value)
```
[the remark this answers is not in the document]
Comment (Neilshweky) [reply lib/mongoid/criteria/queryable/selector.rb:195]: done

File: spec/integration/matcher_examples_spec.rb (hunk 1 above)
Comment (Neilshweky) [reply spec/integration/matcher_examples_spec.rb:693]: I added this test in MONGOID-4911 PR.

**Comment by p-mongo** [comment]
:+1:

**Pull request merged**